quarkchain/evm/state.py

```diff
@@ -92,49 +92,120 @@ class TokenBalancePair(rlp.Serializable):
 
 
 class TokenBalances:
-    """interface for token balances
-    TODODLL: store token balances in trie when TOKEN_TRIE_THRESHOLD is crossed
-    """
+    """Token balances including non-genesis native tokens."""
 
     def __init__(self, data: bytes, db):
-        self.token_trie = SecureTrie(Trie(db))
-        self.balances = {}
-        self.enum = b"\x00"
+        self.token_trie = None
+        self.token_root = None
+        # We don't want outside world to know this
+        self._balances = {}
+        self._db = db
+        self._committed = False
+
         if len(data) != 0:
-            self.enum = data[:1]
-            if self.enum == b"\x00":
+            enum = data[:1]
+            if enum == b"\x00":
                 for p in rlp.decode(data[1:], CountableList(TokenBalancePair)):
-                    self.balances[p.token_id] = p.balance
-            elif self.enum == b"\x01":
-                raise Exception("Token balance trie is not yet implemented")
+                    self._balances[p.token_id] = p.balance
+            elif enum == b"\x01":
+                self.token_root = data[1:]
+                self.token_trie = SecureTrie(Trie(db, data[1:]))
             else:
-                raise Exception("Unknown enum byte in token_balances")
+                raise ValueError("Unknown enum byte in token_balances")
+
+    def commit(self):
+        self._committed = True
```

Comment:
Similar to storage trie, I think we could remove _commit field here and clear _balances since _balances may never be used after commit

Reply:
makes sense. updated.